File: frontend_projects/SmartTavern/src/services/chatBranches.js
// SmartTavern Frontend API Client — Chat Branches (v1)
// 调用后端 modules.smarttavern/chat_branches 系列接口。
// 默认网关：http://localhost:8050/api/modules（见后端 core/config/api_config.py）
//
// 用法：
//   import ChatBranches from '@/services/chatBranches'
//   const latest = await ChatBranches.getLatestMessageByFile('backend_projects/SmartTavern/data/conversations/branch_demo.json')
//   console.log(latest) // { node_id, role, content, depth }
//
// 说明：
// - 按后端契约，所有接口采用 POST + JSON 请求体。
// - CORS 已在网关侧开启。
// - 若需要其它派生接口，可使用 openaiMessages/branchTable。

const DEFAULT_BASE = 'http://localhost:8050/api/modules';

function ensureBase() {
  // 可通过 window.ST_API_BASE 覆盖
  const fromWindow = typeof window !== 'undefined' && window.ST_API_BASE;
  return String(fromWindow || DEFAULT_BASE).replace(/\/+$/, '');
}

async function postJSON(path, body = {}) {
  const base = ensureBase();
  const url = `${base}/${String(path).replace(/^\/+/, '')}`;
  let resp;
  try {
    resp = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body || {}),
    });
  } catch (networkError) {
    const err = new Error(`[ChatBranches] Network error: ${networkError?.message || networkError}`);
    err.cause = networkError;
    err.url = url;
    throw err;
  }

  let data = null;
  const text = await resp.text().catch(() => '');
  try {
    data = text ? JSON.parse(text) : null;
  } catch (parseError) {
    const err = new Error(`[ChatBranches] Invalid JSON response (${resp.status}): ${text?.slice(0, 200)}`);
    err.cause = parseError;
    err.status = resp.status;
    err.url = url;
    throw err;
  }

  if (!resp.ok) {
    const err = new Error(`[ChatBranches] HTTP ${resp.status}: ${data && (data.message || data.error) || 'Unknown error'}`);
    err.status = resp.status;
    err.url = url;
    err.details = data;
    throw err;
  }
  return data;
}

// 轻量缓存（仅本会话内存）
const _mem = new Map();
const _ck = (k) => `cb:${k}`;

const ChatBranches = {
  // 获取某个对话文件的最新消息（依据 active_path）
  // 后端实现参考：[python.function(get_latest_message)](api/modules/SmartTavern/chat_branches/chat_branches.py:130)
  async getLatestMessageByFile(file, { useCache = true } = {}) {
    const key = _ck(`latest:${file}`);
    if (useCache && _mem.has(key)) return _mem.get(key);

    const res = await postJSON('smarttavern/chat_branches/get_latest_message', { file });
    _mem.set(key, res);
    return res;
  },

  // 导出 OpenAI messages（可选）
  // 参考：[python.function(openai_messages)](api/modules/SmartTavern/chat_branches/chat_branches.py:68)
  async openaiMessagesByFile(file) {
    return postJSON('smarttavern/chat_branches/openai_messages', { file });
  },

  // 计算分支情况表（可选）
  // 参考：[python.function(branch_table)](api/modules/SmartTavern/chat_branches/chat_branches.py:98)
  async branchTableByFile(file) {
    return postJSON('smarttavern/chat_branches/branch_table', { file });
  },

  // 创建初始对话（从角色卡 messages[0] 作为根消息），返回三件套路径
  // 后端实现见：[python.function(create_conversation)](api/modules/SmartTavern/chat_branches/chat_branches.py:215)
  async createConversation({
    name,
    description = '',
    type = 'threaded',
    character,
    preset,
    persona,
    regex = null,
    worldbook = null,
  }) {
    return postJSON('smarttavern/chat_branches/create_conversation', {
      name,
      description,
      type,
      character_file: character,
      preset_file: preset,
      persona_file: persona,
      regex_file: regex ?? null,
      worldbook_file: worldbook ?? null,
    });
  },

  // 综合设置管理：读取(action=get)或更新(action=update)对话 settings
  // 使用 file 或 slug 二选一定位
  // 字段：preset(string)、character(string)、persona(string)、regex_rules(array)、world_books(array)
  // 后端实现见：[python.function(settings)](api/modules/SmartTavern/chat_branches/chat_branches.py:282)
  async settings({ action, patch, file, slug }) {
    if (!action || !['get', 'update'].includes(action)) {
      throw new Error('[ChatBranches] settings: action must be "get" or "update"');
    }
    if (action === 'update' && (!patch || typeof patch !== 'object')) {
      throw new Error('[ChatBranches] settings: patch required for action=update');
    }
    const body = { action };
    if (patch) body.patch = patch;
    if (file) body.file = file;
    if (slug) body.slug = slug;
    return postJSON('smarttavern/chat_branches/settings', body);
  },

  // 管理 variables：action=get|set|merge|reset；set/merge 需提供 data 对象
  // 使用 file 或 slug 二选一定位
  // 后端实现见：[python.function(variables)](api/modules/SmartTavern/chat_branches/chat_branches.py:215)
  async variables({ action, data, file, slug }) {
    const body = { action };
    if (data !== undefined) body.data = data;
    if (file) body.file = file;
    if (slug) body.slug = slug;
    return postJSON('smarttavern/chat_branches/variables', body);
  },

  // 追加新消息：创建新节点并更新父节点 children 与 active_path
  // 后端实现见：[python.function(append_message)](api/modules/SmartTavern/chat_branches/chat_branches.py:192)
  async appendMessage({ file, node_id, pid, role, content }) {
    if (!file) {
      throw new Error('[ChatBranches] appendMessage: file is required');
    }
    if (!node_id || !pid || !role || !content) {
      throw new Error('[ChatBranches] appendMessage: node_id, pid, role, content are required');
    }
    return postJSON('smarttavern/chat_branches/append_message', {
      file,
      node_id,
      pid,
      role,
      content
    });
  },

  // 更新消息内容：修改指定节点的 content
  // 后端实现见：[python.function(update_message)](api/modules/SmartTavern/chat_branches/chat_branches.py:138)
  async updateMessage({ file, node_id, content }) {
    if (!file) {
      throw new Error('[ChatBranches] updateMessage: file is required');
    }
    if (!node_id || content === undefined) {
      throw new Error('[ChatBranches] updateMessage: node_id and content are required');
    }
    return postJSON('smarttavern/chat_branches/update_message', {
      file,
      node_id,
      content
    });
  },

  // 修剪消息树：删除指定节点及其所有子孙
  // 后端实现见：[python.function(truncate_after)](api/modules/SmartTavern/chat_branches/chat_branches.py:165)
  async truncateAfter({ file, node_id }) {
    if (!file) {
      throw new Error('[ChatBranches] truncateAfter: file is required');
    }
    if (!node_id) {
      throw new Error('[ChatBranches] truncateAfter: node_id is required');
    }
    return postJSON('smarttavern/chat_branches/truncate_after', {
      file,
      node_id
    });
  },

  // 切换分支：切换 active_path 最后节点到目标分支序号
  // 后端实现见：[python.function(switch_branch)](api/modules/SmartTavern/chat_branches/chat_branches.py:222)
  async switchBranch({ file, target_j }) {
    if (!file) {
      throw new Error('[ChatBranches] switchBranch: file is required');
    }
    if (!target_j || target_j < 1) {
      throw new Error('[ChatBranches] switchBranch: target_j must be >= 1');
    }
    return postJSON('smarttavern/chat_branches/switch_branch', {
      file,
      target_j
    });
  },
};

export default ChatBranches;

File: api/modules/SmartTavern/chat_branches/chat_branches.py
# -*- coding: utf-8 -*-
"""
API 封装层：SmartTavern.chat_branches（无状态版）

只提供基于“单个对话分支树文件（最小结构）”的派生视图能力：
- openai_messages(doc): 从最小分支树文件导出 OpenAI Chat messages
- branch_table(doc): 从最小分支树文件计算分支情况表（含最新层 j/n）

最小分支树文件结构（仅四个字段）：
{
  "roots": ["node_id1", "node_id2", ...],  // 所有根节点ID数组
  "nodes": {
    "node_id": { "pid": "parent_id|null", "role": "system|user|assistant", "content": "..." }
  },
  "children": { "parent_id": ["child_id1","child_id2",...] },   // 可选
  "active_path": ["current_root","...","leafId"]                // 第一个元素是当前使用的根节点
}

注意：
- 本模块不再管理任何对话/会话状态（无 create/append/truncate/switch/export/import/list_* 等接口）
- 外部可直接存储与读取 JSON 文件；此处仅负责计算视图
"""
from typing import Any, Dict
import core
from .impl import (
    openai_messages_from_doc as _openai_messages_from_doc,
    branch_table_from_doc as _branch_table_from_doc,
    get_latest_message_from_doc as _get_latest_message_from_doc,
    update_message_content as _update_message_content,
    truncate_after_node as _truncate_after_node,
    append_new_message as _append_new_message,
    create_conversation_impl as _create_conversation_impl,
    settings_impl as _settings_impl,
    variables_impl as _variables_impl,
)


@core.register_api(
    path="smarttavern/chat_branches/openai_messages",
    name="OpenAI 消息导出（无状态）",
    description="从最小分支树文件导出 OpenAI Chat messages 数组。支持传入 doc（JSON 对象）或 file（文件路径）二选一",
    input_schema={
        "type": "object",
        "properties": {
            "doc": {"type": "object", "additionalProperties": True},
            "file": {"type": "string"}
        },
        "additionalProperties": False,
        "oneOf": [
            {"required": ["doc"]},
            {"required": ["file"]}
        ]
    },
    output_schema={
        "type": "object",
        "properties": {
            "messages": {
                "type": "array",
                "items": {
                    "type": "object",
                    "properties": {"role": {"type": "string"}, "content": {"type": "string"}},
                    "required": ["role", "content"],
                },
            },
            "path": {"type": "array", "items": {"type": "string"}},
        },
        "required": ["messages"],
        "additionalProperties": True,
    },
)
def openai_messages(doc: Dict[str, Any] = None, file: str = None) -> Dict[str, Any]:
    return _openai_messages_from_doc(doc=doc, file=file)


@core.register_api(
    path="smarttavern/chat_branches/branch_table",
    name="分支情况表（无状态）",
    description="从最小分支树文件计算分支情况表（含最新层 j/n）。支持传入 doc（JSON 对象）或 file（文件路径）二选一",
    input_schema={
        "type": "object",
        "properties": {
            "doc": {"type": "object", "additionalProperties": True},
            "file": {"type": "string"}
        },
        "additionalProperties": False,
        "oneOf": [
            {"required": ["doc"]},
            {"required": ["file"]}
        ]
    },
    output_schema={
        "type": "object",
        "properties": {
            "latest": {"type": "object", "additionalProperties": True},
            "levels": {"type": "array", "items": {"type": "object", "additionalProperties": True}},
        },
        "required": ["latest", "levels"],
        "additionalProperties": True,
    },
)
def branch_table(doc: Dict[str, Any] = None, file: str = None) -> Dict[str, Any]:
    return _branch_table_from_doc(doc=doc, file=file)


@core.register_api(
    path="smarttavern/chat_branches/get_latest_message",
    name="获取最后一条消息（无状态）",
    description="根据 active_path 提取最后一条消息。支持传入 doc（JSON 对象）或 file（文件路径）二选一",
    input_schema={
        "type": "object",
        "properties": {
            "doc": {"type": "object", "additionalProperties": True},
            "file": {"type": "string"}
        },
        "additionalProperties": False,
        "oneOf": [
            {"required": ["doc"]},
            {"required": ["file"]}
        ]
    },
    output_schema={
        "type": "object",
        "properties": {
            "node_id": {"type": "string"},
            "role": {"type": "string", "enum": ["system", "user", "assistant"]},
            "content": {"type": "string"},
            "depth": {"type": "integer"}
        },
        "required": ["node_id", "role", "content", "depth"],
        "additionalProperties": False,
    },
)
def get_latest_message(doc: Dict[str, Any] = None, file: str = None) -> Dict[str, Any]:
    return _get_latest_message_from_doc(doc=doc, file=file)


@core.register_api(
    path="smarttavern/chat_branches/update_message",
    name="修改消息内容",
    description="修改指定节点的 content，返回更新后的完整 doc（含 updated_at）",
    input_schema={
        "type": "object",
        "properties": {
            "node_id": {"type": "string"},
            "content": {"type": "string"},
            "doc": {"type": "object", "additionalProperties": True},
            "file": {"type": "string"}
        },
        "required": ["node_id", "content"],
        "additionalProperties": False,
        "oneOf": [
            {"required": ["node_id", "content", "doc"]},
            {"required": ["node_id", "content", "file"]}
        ]
    },
    output_schema={
        "type": "object",
        "additionalProperties": True
    },
)
def update_message(node_id: str, content: str, doc: Dict[str, Any] = None, file: str = None) -> Dict[str, Any]:
    return _update_message_content(node_id=node_id, content=content, doc=doc, file=file)


@core.register_api(
    path="smarttavern/chat_branches/truncate_after",
    name="修剪消息树",
    description="删除指定节点及其所有子孙；更新 nodes/children/active_path（截断到父节点）/updated_at",
    input_schema={
        "type": "object",
        "properties": {
            "node_id": {"type": "string"},
            "doc": {"type": "object", "additionalProperties": True},
            "file": {"type": "string"}
        },
        "required": ["node_id"],
        "additionalProperties": False,
        "oneOf": [
            {"required": ["node_id", "doc"]},
            {"required": ["node_id", "file"]}
        ]
    },
    output_schema={
        "type": "object",
        "additionalProperties": True
    },
)
def truncate_after(node_id: str, doc: Dict[str, Any] = None, file: str = None) -> Dict[str, Any]:
    return _truncate_after_node(node_id=node_id, doc=doc, file=file)


@core.register_api(
    path="smarttavern/chat_branches/append_message",
    name="追加新消息",
    description="创建新节点，更新父节点 children 与 active_path，返回完整 doc（含 updated_at）",
    input_schema={
        "type": "object",
        "properties": {
            "node_id": {"type": "string"},
            "pid": {"type": "string"},
            "role": {"type": "string", "enum": ["system", "user", "assistant"]},
            "content": {"type": "string"},
            "doc": {"type": "object", "additionalProperties": True},
            "file": {"type": "string"}
        },
        "required": ["node_id", "pid", "role", "content"],
        "additionalProperties": False,
        "oneOf": [
            {"required": ["node_id", "pid", "role", "content", "doc"]},
            {"required": ["node_id", "pid", "role", "content", "file"]}
        ]
    },
    output_schema={
        "type": "object",
        "additionalProperties": True
    },
)
def append_message(node_id: str, pid: str, role: str, content: str, doc: Dict[str, Any] = None, file: str = None) -> Dict[str, Any]:
    return _append_new_message(node_id=node_id, pid=pid, role=role, content=content, doc=doc, file=file)


@core.register_api(
    path="smarttavern/chat_branches/create_conversation",
    name="创建初始对话文件（从角色卡 messages[0] 派生）",
    description=(
        "根据 NewGame 选择项创建一套对话三件套：{slug}.json / {slug}.settings.json / {slug}.variables.json。"
        "从角色卡文件读取 messages[0] 作为根消息（暂不启用多分支）。"
    ),
    input_schema={
        "type": "object",
        "properties": {
            "name": {"type": "string"},
            "description": {"type": "string"},
            "type": {"type": "string", "enum": ["threaded"]},
            "character_file": {"type": "string"},
            "preset_file": {"type": "string"},
            "persona_file": {"type": "string"},
            "regex_file": {"type": ["string", "null"]},
            "worldbook_file": {"type": ["string", "null"]},
        },
        "required": ["name", "description", "character_file", "preset_file", "persona_file"],
        "additionalProperties": False,
    },
    output_schema={
        "type": "object",
        "properties": {
            "file": {"type": "string"},
            "settings_file": {"type": "string"},
            "variables_file": {"type": "string"},
            "name": {"type": "string"},
            "root_node_id": {"type": "string"},
            "nodes_count": {"type": "integer"},
            "updated_at": {"type": "string"},
            "slug": {"type": "string"},
        },
        "required": ["file", "settings_file", "variables_file", "name", "root_node_id", "nodes_count", "updated_at", "slug"],
        "additionalProperties": True,
    },
)
def create_conversation(
    name: str,
    description: str,
    character_file: str,
    preset_file: str,
    persona_file: str,
    regex_file: str = None,
    worldbook_file: str = None,
    type: str = "threaded",
) -> Dict[str, Any]:
    return _create_conversation_impl(
        name=name,
        description=description,
        character_file=character_file,
        preset_file=preset_file,
        persona_file=persona_file,
        regex_file=regex_file,
        worldbook_file=worldbook_file,
        chat_type=type or "threaded",
    )


@core.register_api(
    path="smarttavern/chat_branches/settings",
    name="对话设置综合管理（settings.json）",
    description=(
        "读取或更新 conversations/{name}.settings.json。"
        "action=get: 读取当前设置；action=update: 更新指定字段。"
        "允许字段：preset(string)、character(string)、persona(string)、regex_rules(array)、world_books(array)。"
        "使用 file 或 slug 二选一定位。"
    ),
    input_schema={
        "type": "object",
        "properties": {
            "action": {"type": "string", "enum": ["get", "update"]},
            "patch": {
                "type": "object",
                "properties": {
                    "preset": {"type": "string"},
                    "character": {"type": "string"},
                    "persona": {"type": "string"},
                    "regex_rules": {"type": "array", "items": {"type": "string"}},
                    "world_books": {"type": "array", "items": {"type": "string"}}
                },
                "additionalProperties": False
            },
            "file": {"type": "string"},
            "slug": {"type": "string"}
        },
        "required": ["action"],
        "additionalProperties": False,
        "oneOf": [
            {"required": ["action", "file"]},
            {"required": ["action", "slug"]}
        ]
    },
    output_schema={
        "type": "object",
        "properties": {
            "settings_file": {"type": "string"},
            "settings": {"type": "object", "additionalProperties": True},
            "slug": {"type": "string"},
        },
        "required": ["settings_file", "settings", "slug"],
        "additionalProperties": False,
    },
)
def settings(action: str, file: str = None, slug: str = None, patch: Dict[str, Any] = None) -> Dict[str, Any]:
    return _settings_impl(action=action, file=file, slug=slug, patch=patch)


@core.register_api(
    path="smarttavern/chat_branches/variables",
    name="对话变量管理（variables.json）",
    description="管理 conversations/{name}.variables.json：action=get|set|merge|reset。set/merge 需 data 对象。使用 file 或 slug 二选一定位。",
    input_schema={
        "type": "object",
        "properties": {
            "action": {"type": "string", "enum": ["get", "set", "merge", "reset"]},
            "data": {"type": "object"},
            "file": {"type": "string"},
            "slug": {"type": "string"}
        },
        "required": ["action"],
        "additionalProperties": False,
        "oneOf": [
            {"required": ["action", "file"]},
            {"required": ["action", "slug"]}
        ]
    },
    output_schema={
        "type": "object",
        "properties": {
            "variables_file": {"type": "string"},
            "variables": {"type": "object", "additionalProperties": True},
            "slug": {"type": "string"},
        },
        "required": ["variables_file", "variables", "slug"],
        "additionalProperties": False,
    },
)
def variables(action: str, file: str = None, slug: str = None, data: Dict[str, Any] = None) -> Dict[str, Any]:
    return _variables_impl(action=action, file=file, slug=slug, data=data)




